compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -289,6 +289,7 @@ pub enum Visibility<Id = LocalDefId> {
 }
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, HashStable, TyEncodable, TyDecodable)]
+#[derive(TypeFoldable, TypeVisitable)]
```

Comment:
This one's a bit annoying actually, as it's *solely* in order to satisfy the following chain (at the root of which is just a debug assertion):

-  `Binder::rebind<U>` requires that `U: TypeVisitable` so that (only when debug assertions are enabled) bound vars can be validated;
- `<Match as TypeRelation>::binders` calls the above with generic `U: Relate`;
- `Relate` therefore has `TypeVisitable` as a supertrait (actually it needlessly has `TypeFoldable`, which itself is a subtrait of `TypeVisitable`);
- `<TraitPredicate as Relate>::relate` establishes the relation between values of its `constness` field by via `<BoundConstness as Relate>::relate`; and
- `BoundConstness` therefore impls `Relate` and so is required to impl the visitable traits.

I'm not sure it's worth changing... but is it worth detailing all of the above in a comment?

Reply:
Ah, actually `Relate: TypeFoldable` is actually required elsewhere further downstream.

Reply:
Added explanation in 8834fbf4346d25f3051044979379897299ca5e7f